### PR TITLE
Add interactive textual client

### PR DESCRIPTION
**Pull request by frcroth**
Supports listing keys and getting keys. There is a delete button, which I did not implement because I did not want to destroy my annotations (but should be easily done).

~~Edit: CI doesn't work because of ancient python version~~

### CODE REVIEW

**Review by fm3 — COMMENTED** [review]
This is a great new tool in our toolbox! Making me happy!

It already works quite well. Would it be hard to add the following features from my wish-list?

 - Add a button similar to k to show the *previous* keys? Right now, I can only navigate in one direction
 - Instead of these two buttons, have the arrow keys up/down move to the next keys when the end/start of the current list is reached?
 - Make the fossildb address configurable (but default to localhost:7155)? This could also be a command line argument
 - Have the output files be named after the key and the version? e.g. <key>_v<version> rather than out.bin
 - Remove the delete button for now (I think delete is not a common use case, and it does not work now anyway)

For the moment, let’s assess the effort required for these features, and not start right away with implementing. We can then discuss the priority.

**Comment by frcroth** [comment]
>     * Add a button similar to k to show the _previous_ keys? Right now, I can only navigate in one direction

With a list of previous offsets that should be relatively easy.

>     * Instead of these two buttons, have the arrow keys up/down move to the next keys when the end/start of the current list is reached?

Would need to take a look at how the list works. Most cli applications have 2 ways of navigating, page up/down and line by line. The arrow down (or j and k in vi) should only scroll one line.
 
>     * Make the fossildb address configurable (but default to localhost:7155)? This could also be a command line argument

Via file trivial, don't know immediately know how to do it via cli because we don't really have control of the startup procedure. But should be a common use case and not too hard.
 
>     * Have the output files be named after the key and the version? e.g. _v rather than out.bin

Seems very straight forward. Not clear what naming scheme you're proposing: <key_name>_v ? 
 
>     * Remove the delete button for now (I think delete is not a common use case, and it does not work now anyway)

Well that's easy.
 
 
> For the moment, let’s assess the effort required for these features, and not start right away with implementing. We can then discuss the priority.

**Comment by fm3** [comment]
Thanks for your thoughts! I think we should go for these ideas. On output file naming: sorry, markup ate my text. I meant `<key>_v<version>`

I think having a list of previous to scroll back is fair :) Might lead to slightly wrong offsets if the server content has changed, but I think that should not do much harm.

On command line arguments: I would guess that you could do argparse before the `app = FossilDBClient(stub)` line, and pass the result as a second parameter there.

If changing the scrolling behavior is hard, Let’s make it a follow-up issue. Maybe an easier improvement can be found also. My worry is that users won’t notice at all that there might be more keys than are shown (especially since the list does not fill the whole widget/screen if the terminal is larger). Maybe the presence of more keys can be detected (by requesting slightly more than are shown?) and in this case add some sort of marker in the list? Also, could page-down and page-up maybe replace the k shortcut? But again, all of this can be a follow-up issue.

Could you address the other points?

File: client/.gitignore [hunk 1]
```diff
@@ -3,3 +3,5 @@ __pycache__
 
 fossildbapi_pb2_grpc.py
 fossildbapi_pb2.py
+
+env
```
Comment (fm3) [inline client/.gitignore:7]:
```suggestion
env

```

File: client/interactive/client.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+
+# import logging
+
+# logFile = "sample.log"
+# logging.basicConfig(
+#     filename=logFile,
+#     filemode="a",
+#     level=logging.DEBUG,
+#     format="%(asctime)s - %(levelname)s: %(message)s",
+# )
```
Comment (fm3) [inline client/interactive/client.py:23]:
```suggestion
```

File: client/interactive/client.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    def on_data_table_row_highlighted(self, event):
+        self.query_one(KeyInfoWidget).update_key(event.value)
+        self.refresh()
+
+
+class FileNameHint(Widget):
+    filename = reactive("none")
+
+    def render(self) -> str:
+        return f"Press the button above to save the value at the key in {self.filename}"
```
Comment (fm3) [inline client/interactive/client.py:49]:
```suggestion
        return f"Press the button above to download the value for the selected key to {self.filename}"
```

File: client/interactive/client.tcss [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    width: 30%;
+}
+
+#key-info-label {
+    height: 50%;
+}
+
+Button {
+    width: 100%;
+}
```
Comment (fm3) [inline client/interactive/client.tcss:16]:
```suggestion
}

```

File: client/update_api.sh [hunk 5]
```diff
@@ -4,3 +4,4 @@ set -Eeuo pipefail
 SCRIPTPATH="$( cd "$(dirname "$0")" ; pwd -P )"
 
 python3 -m grpc_tools.protoc -I$SCRIPTPATH/../src/main/protobuf --python_out=$SCRIPTPATH --grpc_python_out=$SCRIPTPATH $SCRIPTPATH/../src/main/protobuf/fossildbapi.proto
+cp $SCRIPTPATH/fossildbapi_pb2* $SCRIPTPATH/interactive/
```
Comment (fm3) [inline client/update_api.sh:7]:
```suggestion
cp $SCRIPTPATH/fossildbapi_pb2* $SCRIPTPATH/interactive/

```

**Review by fm3 — APPROVED** [review]
:tada:

**Pull request merged**